backend/controller/scaling/local_scaling.go

```diff
@@ -0,0 +1,116 @@
+package scaling
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/alecthomas/errors"
+	"github.com/alecthomas/kong"
+
+	"github.com/TBD54566975/ftl/backend/common/log"
+	"github.com/TBD54566975/ftl/backend/common/model"
+	"github.com/TBD54566975/ftl/backend/runner"
+)
+
+var _ RunnerScaling = (*LocalScaling)(nil)
+
+type LocalScaling struct {
+	cacheDir string
+	runners  []model.RunnerKey
+
+	nextBind            *url.URL
+	controllerAddresses []*url.URL
+}
+
+func NewLocalScaling(nextBind *url.URL, controllerAddresses []*url.URL) (*LocalScaling, error) {
+	cacheDir, err := os.UserCacheDir()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return &LocalScaling{
+		cacheDir:            cacheDir,
+		nextBind:            nextBind,
+		controllerAddresses: controllerAddresses,
+	}, nil
+}
+
+func (l *LocalScaling) SetReplicas(ctx context.Context, replicas int) error {
+	logger := log.FromContext(ctx)
+	logger.Infof("SetReplicas: %d", replicas)
+
+	replicasToAdd := replicas - len(l.runners)
+
+	logger.Infof("Requested replicas %d", replicas)
+	if replicasToAdd <= 0 {
+		return nil
+	}
+
+	logger.Infof("Adding %d replicas", replicasToAdd)
+	for i := 0; i < replicasToAdd; i++ {
+		i := i
+
+		var err error
+		l.nextBind, err = IncrementPort(l.nextBind)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		controllerEndpoint := l.controllerAddresses[i%len(l.controllerAddresses)]
+		config := runner.Config{
+			Bind:               l.nextBind,
+			ControllerEndpoint: controllerEndpoint,
+		}
+
+		name := fmt.Sprintf("runner%d", i)
+		if err := kong.ApplyDefaults(&config, kong.Vars{
+			"deploymentdir": filepath.Join(l.cacheDir, "ftl-runner", name, "deployments"),
+			"language":      "go,kotlin",
+		}); err != nil {
+			return errors.WithStack(err)
+		}
+
+		// Create a readable ULID for the runner.
+		var ulid [16]byte
+		binary.BigEndian.PutUint32(ulid[10:], uint32(len(l.runners)+1))
+		ulidStr := fmt.Sprintf("%025X", ulid)
+		err = config.Key.Scan(ulidStr)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		runnerCtx := log.ContextWithLogger(ctx, logger.Scope(name))
+		l.runners = append(l.runners, config.Key)
```

Comment:
Mutations to this will need to be behind a lock, because multiple controllers could be calling this method concurrently.